base/compiler/ssair/passes.jl

```diff
@@ -1128,23 +1133,28 @@ function try_resolve_finalizer!(ir::IRCode, idx::Int, finalizer_idx::Int, defuse
     all(check_in_range, defuse.defs) || return nothing
 
     # For now: Require all statements in the basic block range to be nothrow.
-    all(minval:maxval) do idx::Int
-        return is_nothrow(ir, idx) || idx == finalizer_idx
+    all(minval:maxval) do sidx::Int
+        return is_nothrow(ir, idx) || sidx == finalizer_idx || sidx == idx
     end || return nothing
 
     # Ok, `finalizer` rewrite is legal.
     finalizer_stmt = ir[SSAValue(finalizer_idx)][:inst]
     argexprs = Any[finalizer_stmt.args[2], finalizer_stmt.args[3]]
-    inline = finalizer_stmt.args[4]
-    if inline === nothing
-        # No code in the function - Nothing to do
-    else
-        mi = finalizer_stmt.args[5]::MethodInstance
-        if inline::Bool && try_inline_finalizer!(ir, argexprs, maxval, mi, inlining)
-            # the finalizer body has been inlined
+    flags = info === nothing ? UInt8(0) : flags_for_effects(info.effects)
+    if length(finalizer_stmt.args) >= 4
+        inline = finalizer_stmt.args[4]
+        if inline === nothing
+            # No code in the function - Nothing to do
         else
-            insert_node!(ir, maxval, NewInstruction(Expr(:invoke, mi, argexprs...), Nothing), true)
+            mi = finalizer_stmt.args[5]::MethodInstance
+            if inline::Bool && try_inline_finalizer!(ir, argexprs, maxval, mi, inlining)
+                # the finalizer body has been inlined
+            else
+                insert_node!(ir, maxval, with_flags(NewInstruction(Expr(:invoke, mi, argexprs...), Nothing), flags), true)
+            end
         end
+    else
+        insert_node!(ir, maxval, with_flags(NewInstruction(Expr(:call, argexprs...), Nothing), flags), true)
```

Comment:
So is my understanding right that this branch is for generating `Core.finalizer` call expression that may be handled by the second run of this optimization pass?